Mathlib/Data/List/Basic.lean

```diff
@@ -785,8 +801,31 @@ theorem getLast?_append {l₁ l₂ : List α} {x : α} (h : x ∈ l₂.getLast?)
     exact h
 #align list.last'_append List.getLast?_append
 
+@[simp]
+theorem getLast?_reverse (l : List α) : l.reverse.getLast? = l.head? := by cases l <;> simp
+
+/-! ### getLast! -/
+
+theorem getLast!_eq_getLast_of_ne_nil [Inhabited α] {l : List α} (h : l ≠ []) :
+    l.getLast! = l.getLast h := by
+  match l with
+  | [] => contradiction
+  | a :: as => rfl
+
+@[simp]
+theorem getLast!_reverse [Inhabited α] (l : List α) : l.reverse.getLast! = l.head! := by
+  match l with
+  | [] => rfl
+  | a :: as => simp [getLast!_eq_getLast_of_ne_nil]; rfl
+
 /-! ### head(!?) and tail -/
 
+theorem get_zero_eq_head_of_ne_nil {l : List α} (h : l ≠ []) :
+    l.get ⟨ 0, length_pos_iff_ne_nil.mpr h ⟩ = l.head h := by
```

Comment:
AFAICT, this can't be used in `@[simp]` because it can't match `h` based on LHS. See also https://github.com/leanprover/std4/pull/758